// @aws-cdk/cloud-assembly-schema/lib/manifest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Manifest = exports.VALIDATION_REPORT_FILE = exports.VERSION_MISMATCH = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const fs = __importStar(require("fs"));
const jsonschema = __importStar(require("jsonschema"));
const semver = __importStar(require("semver"));
const assembly = __importStar(require("./cloud-assembly"));
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
// this prefix is used by the CLI to identify this specific error.
// in which case we want to instruct the user to upgrade his CLI.
// see exec.ts#createAssembly
exports.VERSION_MISMATCH = 'Cloud assembly schema version mismatch';
/** Canonical filename of the policy-validation report emitted into a cloud assembly. */
exports.VALIDATION_REPORT_FILE = 'validation-report.json';
/**
 * CLI version is created at build and release time
 *
 * It needs to be .gitignore'd, otherwise the projen 'no uncommitted
 * changes' self-check will fail, which means it needs to be generated
 * at build time if it doesn't already exist.
 */
const CLI_VERSION = require("../cli-version.json");
const ASSETS_SCHEMA = require("../schema/assets.schema.json");
const ASSEMBLY_SCHEMA = require("../schema/cloud-assembly.schema.json");
const INTEG_SCHEMA = require("../schema/integ.schema.json");
const VALIDATION_REPORT_SCHEMA = require("../schema/validation-report.schema.json");
/**
 * Version is shared for both manifests
 */
const SCHEMA_VERSION = require("../schema/version.json");
/**
 * Protocol utility class.
 */
class Manifest {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/cloud-assembly-schema.Manifest", version: "54.12.0" };
    /**
     * Validates and saves the cloud assembly manifest to file.
     *
     * @param manifest - manifest.
     * @param filePath - output file path.
     */
    static saveAssemblyManifest(manifest, filePath) {
        Manifest.saveManifest(manifest, filePath, ASSEMBLY_SCHEMA);
    }
    /**
     * Load and validates the cloud assembly manifest from file.
     *
     * @param filePath - path to the manifest file.
     */
    static loadAssemblyManifest(filePath, options) {
        return Manifest.loadManifest(filePath, ASSEMBLY_SCHEMA, Manifest.patchStackTagsOnRead, options);
    }
    /**
     * Validates and saves the asset manifest to file.
     *
     * @param manifest - manifest.
     * @param filePath - output file path.
     */
    static saveAssetManifest(manifest, filePath) {
        Manifest.saveManifest(manifest, filePath, ASSETS_SCHEMA);
    }
    /**
     * Load and validates the asset manifest from file.
     *
     * @param filePath - path to the manifest file.
     */
    static loadAssetManifest(filePath) {
        return Manifest.loadManifest(filePath, ASSETS_SCHEMA);
    }
    /**
     * Validates and saves the integ manifest to file.
     *
     * @param manifest - manifest.
     * @param filePath - output file path.
     */
    static saveIntegManifest(manifest, filePath) {
        Manifest.saveManifest(manifest, filePath, INTEG_SCHEMA);
    }
    /**
     * Load and validates the integ manifest from file.
     *
     * @param filePath - path to the manifest file.
     */
    static loadIntegManifest(filePath) {
        const manifest = this.loadManifest(filePath, INTEG_SCHEMA);
        // Adding typing to `validate()` led to `loadManifest()` to properly infer
        // its return type, which indicated that the return type of this
        // function may be a lie. I could change the schema to make `testCases`
        // optional, but that will bump the major version of this package and I
        // don't want to do that. So instead, just make sure `testCases` is always there.
        return {
            ...manifest,
            testCases: manifest.testCases ?? [],
        };
    }
    /**
     * Load and validate the policy validation report from file.
     *
     * @param filePath - path to the validation report file.
     */
    static loadValidationReport(filePath) {
        return Manifest.loadManifest(filePath, VALIDATION_REPORT_SCHEMA);
    }
    /**
     * Fetch the current schema version number.
     */
    static version() {
        return `${SCHEMA_VERSION.revision}.0.0`;
    }
    /**
     * Return the CLI version that supports this Cloud Assembly Schema version
     */
    static cliVersion() {
        const version = CLI_VERSION.version;
        return version ? version : undefined;
    }
    /**
     * Deprecated
     * @deprecated use `saveAssemblyManifest()`
     */
    static save(manifest, filePath) {
        return this.saveAssemblyManifest(manifest, filePath);
    }
    /**
     * Deprecated
     * @deprecated use `loadAssemblyManifest()`
     */
    static load(filePath) {
        return this.loadAssemblyManifest(filePath);
    }
    static validate(manifest, schema, options) {
        function parseVersion(version) {
            const ver = semver.valid(version);
            if (!ver) {
                throw new Error(`Invalid semver string: "${version}"`);
            }
            return ver;
        }
        const maxSupported = semver.major(parseVersion(Manifest.version()));
        const actual = parseVersion(manifest.version);
        // first validate the version should be accepted. all versions within the same minor version are fine
        if (maxSupported < semver.major(actual) && !options?.skipVersionCheck) {
            // If we have a more specific error to throw than the generic one below, make sure to add that info.
            const cliVersion = manifest.minimumCliVersion;
            let cliWarning = '';
            if (cliVersion) {
                cliWarning = `. You need at least CLI version ${cliVersion} to read this manifest.`;
            }
            // we use a well known error prefix so that the CLI can identify this specific error
            // and print some more context to the user.
            throw new Error(`${exports.VERSION_MISMATCH}: Maximum schema version supported is ${maxSupported}.x.x, but found ${actual}${cliWarning}`);
        }
        if (options?.validateSchema ?? (process.env.TESTING_CDK === '1')) {
            // now validate the format is good.
            const validator = new jsonschema.Validator();
            const result = validator.validate(manifest, schema, {
                nestedErrors: true,
                allowUnknownAttributes: false,
                preValidateProperty: Manifest.validateAssumeRoleAdditionalOptions,
            });
            let errors = result.errors;
            if (options?.skipEnumCheck) {
                // Enum validations aren't useful when
                errors = stripEnumErrors(errors);
            }
            if (errors.length > 0) {
                throw new Error(`Invalid assembly manifest:\n${errors.map((e) => e.stack).join('\n')}`);
            }
        }
    }
    static saveManifest(manifest, filePath, schema, preprocess) {
        let withVersion = {
            ...manifest,
            version: Manifest.version(),
            minimumCliVersion: Manifest.cliVersion(),
        };
        Manifest.validate(withVersion, schema);
        if (preprocess) {
            withVersion = preprocess(withVersion);
        }
        fs.writeFileSync(filePath, JSON.stringify(withVersion, undefined, 2));
    }
    static loadManifest(filePath, schema, preprocess, options) {
        const contents = fs.readFileSync(filePath, { encoding: 'utf-8' });
        let obj;
        try {
            obj = JSON.parse(contents);
        }
        catch (e) {
            throw new Error(`${e.message}, while parsing ${JSON.stringify(contents)}`);
        }
        if (preprocess) {
            obj = preprocess(obj);
        }
        Manifest.validate(obj, schema, options);
        return obj;
    }
    /**
     * Fix the casing of stack tags entries
     *
     * At the very beginning of the CDK we used to emit stack tags as an object with
     * `{ Key, Value }` keys; this had the "advantage" that we could stick those
     * tags directly into the `CreateChangeSet` call.
     *
     * Then we later on used jsii on the assembly schema and we were forced to type
     * the in-memory objects as `{ key, value }` with lowercase letters. Now the
     * objects have a different on-disk and in-memory format, and we need to convert
     * between them.
     *
     * For backwards compatibility reasons, we used to convert lowercase in-memory
     * to uppercase on-disk variant until very recently. This is now unnecessary,
     * since no officially supported CDK tools read the stack tags from the
     * metadata; the CLI and toolkit library read stack tags from the artifact
     * properties.
     *
     * So although we don't emit uppercase stack tag objects anymore, we might still read
     * manifests that have them. Because the manifest we read must pass JSON Schema
     * validation (which expects lowercase tag objects), we have to fix the casing
     * of these objects after reading from disk and before validating.
     *
     * That's what this function does.
     */
    static patchStackTagsOnRead(manifest) {
        const artifacts = Object.values(manifest.artifacts ?? {})
            .filter(artifact => artifact.type === assembly.ArtifactType.AWS_CLOUDFORMATION_STACK);
        for (const artifact of artifacts) {
            const tagMetadata = Object.values(artifact.metadata ?? {})
                .flatMap(x => x)
                .filter(entry => entry.type === assembly.ArtifactMetadataEntryType.STACK_TAGS);
            for (const entry of tagMetadata) {
                const tags = entry.data;
                for (const tag of tags ?? []) {
                    const t = tag;
                    if ('Key' in t) {
                        t.key = t.Key;
                        delete t.Key;
                    }
                    if ('Value' in t) {
                        t.value = t.Value;
                        delete t.Value;
                    }
                }
            }
        }
        return manifest;
    }
    /**
     * Validates that `assumeRoleAdditionalOptions` doesn't contain nor `ExternalId` neither `RoleArn`, as they
     * should have dedicated properties preceding this (e.g `assumeRoleArn` and `assumeRoleExternalId`).
     */
    static validateAssumeRoleAdditionalOptions(instance, key, _schema, _options, _ctx) {
        if (key !== 'assumeRoleAdditionalOptions') {
            // note that this means that if we happen to have a property named like this, but that
            // does want to allow 'RoleArn' or 'ExternalId', this code will have to change to consider the full schema path.
            // I decided to make this less granular for now on purpose because it fits our needs and avoids having messy
            // validation logic due to various schema paths.
            return;
        }
        const assumeRoleOptions = instance[key];
        if (assumeRoleOptions?.RoleArn) {
            throw new Error(`RoleArn is not allowed inside '${key}'`);
        }
        if (assumeRoleOptions?.ExternalId) {
            throw new Error(`ExternalId is not allowed inside '${key}'`);
        }
    }
}
exports.Manifest = Manifest;
function stripEnumErrors(errors) {
    return errors.filter((e) => typeof e.schema === 'string' || !('enum' in e.schema));
}
